server/src/main/java/org/elasticsearch/indices/IndicesService.java

```diff
@@ -740,13 +721,38 @@ private synchronized IndexService createIndexService(
             mapperRegistry,
             indicesFieldDataCache,
             namedWriteableRegistry,
-            idFieldMappers.apply(idxSettings.getMode()),
+            idFieldMappers.apply(indexModule.indexSettings().getMode()),
             valuesSourceRegistry,
             indexFoldersDeletionListeners,
             snapshotCommitSuppliers
         );
     }
 
+    private IndexModule createIndexModule(IndexService.IndexCreationContext indexCreationContext, IndexMetadata indexMetadata) {
+        try (var ignored = threadPool.getThreadContext().newStoredContext()) {
```

Comment:
Hmm, although we do use this within a cluster state applier, I just remembered that we also use it to validate things before creating an index. It's important that deprecation warnings from this validation process do propagate back to the caller.